compiler/z/codegen/S390Peephole.cpp

```diff
@@ -0,0 +1,3248 @@
+         {
+         case TR::InstOpCode::L:
+            {
+            LoadStoreReduction(TR::InstOpCode::ST, 4);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         case TR::InstOpCode::LFH:
+            {
+            LoadStoreReduction(TR::InstOpCode::STFH, 4);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         case TR::InstOpCode::LG:
+            {
+            LoadStoreReduction(TR::InstOpCode::STG, 8);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         default:
+            {
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         }
+
+      moveInstr = true;
+      if(moveInstr)
+         _cursor = _cursor->getNext();
+      }
+
+   if (comp()->getOption(TR_TraceCG))
+      printInfo("\n\n");
+   }
+
+bool
+TR_S390PreRAPeephole::LoadStoreReduction(TR::InstOpCode::Mnemonic storeOpCode, uint16_t size)
+   {
+   if (_cursor->getNext()->getOpCodeValue() == storeOpCode)
+      {
+      TR::S390RXInstruction* loadInst = static_cast<TR::S390RXInstruction*> (_cursor);
+      TR::S390RXInstruction* storeInst = static_cast<TR::S390RXInstruction*> (_cursor->getNext());
+
+      if ( loadInst->getMemoryReference()->getSymbolReference() && loadInst->getMemoryReference()->getSymbolReference()->isUnresolved() ||
+          storeInst->getMemoryReference()->getSymbolReference() && storeInst->getMemoryReference()->getSymbolReference()->isUnresolved())
+         {
+         return false;
+         }
+
+      if (!loadInst->getMemoryReference()->getBaseRegister() && !loadInst->getMemoryReference()->getIndexRegister() ||
+          !storeInst->getMemoryReference()->getBaseRegister() && !storeInst->getMemoryReference()->getIndexRegister())
+         {
+         return false;
+         }
+
+      TR::Register* reg = loadInst->getRegisterOperand(1);
+
+      if (reg != storeInst->getRegisterOperand(1))
+         {
+         return false;
+         }
+
+      // Register must only be used in the load-store sequence, possibly used in the memory reference
+      // of the load however, not in the store since the the register would have to be loaded first.
+      ncount_t uses = 2;
+      uses += reg == loadInst->getMemoryReference()->getBaseRegister();
+
+      // Bail out if register could have future uses, or if either instruction has dependencies.
+      // Since 2 instructions are replaced with one before RA, we can't handle things like merging
+      // dependencies.
+      if (reg->getTotalUseCount() != uses || reg->isDependencySet() ||
+         loadInst->getDependencyConditions() || storeInst->getDependencyConditions())
```

Comment:
This is the first time I'm seeing a use of `isDependencySet`. Why do we need this check? Is it not sufficient to just check whether `loadInst` and `storeInst` have dependencies on them? By the way a register being a part of a dependency list will increment its total use count so we cannot have the case where `reg->getTotalUseCount() == uses && reg->isDependencySet()` is true.

Reply:
76e39d2